Add `*pty` API

DMaroo opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

Add API to access `*pty` functions (`openpty`, `login_tty` and `forkpty`)

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

Add unit tests, green CI

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

Closes #3212

DMaroo committed:
Initial function signatur and struct for `RzPty` API

DMaroo committed:
Change type annotation to borrow

DMaroo committed:
Implement `rz_subprocess_login_tty`

DMaroo committed:
Add fork mode enum and move pty struct; restructure ifdef guards

DMaroo committed:
Add rough implementation of forkpty

  * Add `rz_sys` versions for `forkpty` family of functions

DMaroo committed:
Add forkpty code; Make changes according to review

DMaroo committed:
Implement forkpty properly and add unit test

  * Add `test_interactive_pty`

DMaroo committed:
Add documentation

DMaroo committed:
Remove const in function argument

  * Apparently, MacOS as non-const arguments for `openpty` and `forkpty`

DMaroo committed:
Add more tests for subprocess forkin using forkpty

  * `test_interactive_pipe_pty`: Create pipes between the slave's PTY
    and master and use them
  * `test_interactive_custom_pty`: Use a custom PTY for the slave

DMaroo committed:
Add PTY functions exist checks for unit tests

DMaroo commented:
I need help in resolving the failing tests for MacOS and Windows. It would be really helpful if anyone could help in figuring out what exactly is the cause for the failing CI.

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -967,13 +969,54 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
 	child_env = create_child_env(opt->envvars, opt->envvals, opt->env_size);
 
-	proc->pid = rz_sys_fork();
+	void *term_params = NULL;
+#if HAVE_FORKPTY && HAVE_OPENPTY && HAVE_LOGIN_TTY
+	struct termios t;
+	/* Needed to avoid reading back the writes again from the TTY */
+	tcgetattr(STDIN_FILENO, &t);
+	cfmakeraw(&t);
+	term_params = &t;
+#endif
```
Isn't this necessary only when RZ_SUBPROCESS_FORKPTY?

DMaroo replied:
Yes, indeed. I'll make the change.

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -967,13 +969,54 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+	if (opt->fork_mode == RZ_SUBPROCESS_FORKPTY) {
+		proc->stderr_fd = (proc->stderr_fd == -1) ? master_fd : proc->stderr_fd;
+		proc->stdout_fd = (proc->stdout_fd == -1) ? master_fd : proc->stdout_fd;
+		proc->stdin_fd = (proc->stdin_fd == -1) ? master_fd : proc->stdin_fd;
```
you could just move these in the right branch of the `if` above

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -967,13 +969,54 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+		if (opt->fork_mode == RZ_SUBPROCESS_FORKPTY) {
+			if (opt->pty && !rz_subprocess_login_tty(opt->pty)) {
+				RZ_LOG_ERROR("login_tty failed, falling back to normal forking\n");
+				goto normal;
```
why the `goto`?

DMaroo replied:
Sorry, it is a remnant of some intermediate code I was trying. Removing it, thanks!

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -1096,6 +1141,13 @@ static RzSubprocessWaitReason subprocess_wait(RzSubprocess *proc, ut64 timeout_m
+	/* Check if stdout and stderr are connected to a PTY
+	If yes, then set true if we n_bytes == 0,
+	i.e. waiing for subprocess to end */
+	bool stdout_pty = proc->stdout_fd == proc->master_fd && n_bytes == 0;
+	bool stderr_pty = proc->stderr_fd == proc->master_fd && n_bytes == 0;
```
Is `proc->master_fd` ever initialized when the subprocess is not in a new PTY?

ret2libc replied:
Also, why are we interested in whether stdout/stederr are pty only when n_bytes == 0?

DMaroo replied:
1. `proc->master_fd` is set to -1 for when the subprocess is not in a new PTY.
2. Because `n_bytes == 0` only when we are waiting for the process to die (`n_bytes` are 0, when the caller is `rz_subprocess_wait`). I'll add a comment explaining this in the code.

DMaroo replied:
This can be considered a sort of a hack-ish way to figure out whether we are waiting for the subprocess to die or not. The significance of `n_bytes` being zero is also stated in the function's (`subprocess_wait`) Doxygen:

```
 * ... If \p n_bytes is
 * not 0, the function terminates as soon as \p n_bytes bytes have been read or
 * the timeout expires.
```

ret2libc replied:
Wouldn't this work well also without the `n_bytes == 0`? IN `read_to_strbuf` `read` would return EIO only when EOF.

DMaroo replied:
Fair enough.

ret2libc commented on test/unit/test_subprocess.c:
```diff
@@ -263,6 +263,120 @@ bool test_interactive(void) {
+
+	RzSubprocessOpt opt = { 0 };
+	opt.file = exe_path;
+	opt.stdin_pipe = RZ_SUBPROCESS_PIPE_CREATE;
+	opt.stdout_pipe = RZ_SUBPROCESS_PIPE_CREATE;
+	opt.stderr_pipe = RZ_SUBPROCESS_PIPE_CREATE;
+	opt.fork_mode = RZ_SUBPROCESS_FORKPTY;
```
I don't remember anymore, but what is the diff between PIPE_NONE + FORKPTY and PIPE_CREATE + PTY? In other words, what is the diff between this test and test_interactive_pty?

DMaroo replied:
In the test `test_interactive_pty`, the subprocess reads its `stdin` and writes its `stdout` to a single slave file descriptor, and the parent process reads the `stdout` and writes the `stdin` of the child to a single master file descriptor.

In the test `test_interactive_pipe_pty`, the subprocess reads its `stdin` and writes its `stdout` to a single slave file descriptor (this stays the same), however, there are two pipes, connecting the master file descriptor to two distinct file descriptors, one for the `stdin` and other for the `stdout`. Hence the parent process reads the `stdout` and writes the `stdin` of the child to two distinct file descriptors, both of which belong to pipes, whose other ends point to the master file descriptor.

DMaroo replied:
But currently, `test_interactive_pipe_pty` passes but doesn't do what I specified above. It directly connects to the slave file descriptor. I'm working on changing this.

DMaroo replied:
Upon further thinking, I can't seem to think of a use case where using pipes with forkpty would be any useful. I will just not use pipes if the `fork_mode` is set to `RZ_SUBPROCESS_FORKPTY`. This means that there is no need for the `test_interactive_pipe_pty` test. I'll remove that.

Please let me know if there are any issues with it. Thanks!

ret2libc reviewed (commented):
Probably the failures are caused by changes in RzSubprocess?

DMaroo committed:
Move terminal param code inside the forkpty branch

DMaroo committed:
Add comment about why we need to check `n_bytes = 0` in `subprocess_wait`

DMaroo commented:
Also, I don't think we can support invoking `rz_subprocess_start_opt` with an existing `RzPTY`, because the logic in `subprocess_wait` assumes contiguous file descriptors. To pass a PTY, we would need to create one before hand, henceforth breaking this assumption. Fixing this would require major changes in `subprocess_wait`, hence I am dropping the support for passing a `RzPTY`. I don't think there would be a scenario where it would be particularly useful.

DMaroo committed:
Remove forkpty with a specified `RzPTY`

  * Also, no piping done in forkpty

ret2libc commented:
> Also, I don't think we can support invoking `rz_subprocess_start_opt` with an existing `RzPTY`, because the logic in `subprocess_wait` assumes contiguous file descriptors. To pass a PTY, we would need to create one before hand, henceforth breaking this assumption. Fixing this would require major changes in `subprocess_wait`, hence I am dropping the support for passing a `RzPTY`. I don't think there would be a scenario where it would be particularly useful.

Right now I don't remember exactly if that would be useful, but for ideas on how we plan to use RzSubprocess more look at run.c:handle_redirection_proc. A lot of the code in run.c should be replaced with the more uniform and standard RzSubprocess interface.

BTW, where does `subprocess_wait` assume that the files are contiguous?

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -927,6 +929,13 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+
+	/* There seems to be no reasonable use case for having pipes when forking
+	using forkpty, hence we will bypass all the pipe creation logic */
+	if (opt->fork_mode == RZ_SUBPROCESS_FORKPTY) {
```
instead of `goto`, extract the pipe handling in a separate function, like

```
if (opt->for_mode != RZ_SUBPROCESS_FORKPTY && !start_handle_pipes(proc, opt)) {
   goto error;
}
```

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -963,14 +972,38 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+		proc->pid = rz_sys_forkpty(&master_fd, NULL, term_params, NULL);
+		proc->master_fd = master_fd;
+
+		proc->stderr_fd = (proc->stderr_fd == -1) ? master_fd : proc->stderr_fd;
+		proc->stdout_fd = (proc->stdout_fd == -1) ? master_fd : proc->stdout_fd;
+		proc->stdin_fd = (proc->stdin_fd == -1) ? master_fd : proc->stdin_fd;
```
I guess stderr_fd and the others are always -1 if we are in this branch. Maybe just put a `rz_warn_if_fail(proc->stderr_fd == 1);` etc. just to be sure, but it should never happen

ret2libc commented on test/unit/test_subprocess.c:
```diff
@@ -263,6 +263,73 @@ bool test_interactive(void) {
+	RzSubprocessOpt opt = { 0 };
+	opt.file = exe_path;
+	opt.stdin_pipe = RZ_SUBPROCESS_PIPE_NONE;
+	/* These won't matter since in forkpty no pipes are created */
+	opt.stdout_pipe = RZ_SUBPROCESS_PIPE_CREATE;
+	opt.stderr_pipe = RZ_SUBPROCESS_PIPE_CREATE;
```
I would prefer to have the API enforce that these are PIPE_NONE and fail if they aren't, otherwise it is not clear what a user can or cannot do.

DMaroo committed:
Add checks to disallow using `forkpty` with piping

DMaroo committed:
Make changes according to review

  * No need to check proc file descriptors for -1 in `forkpty` mode, since
    they *will* be -1
  * No need to check `n_bytes == 0` in for `std{out,err}_pty`

DMaroo committed:
Extract pipe initialization code in `init_pipes` function

DMaroo commented:
I'm working on making `subprocess_wait` work with an existing `RzPTY` object.

DMaroo committed:
Allow using custom PTY for `rz_subprocess_forkpty`

DMaroo commented:
Now, we can use an existing `RzPTY` object as the master-slave pair for the PTY in `rz_subprocess_forkpty`. The test `test_interactive_custom_pty` in `test_subprocess.c` tests this.

ret2libc commented:
MacOS ARM does not fail for me locally, however I see it is failing in the CI. Also BSD :( Unsure what could be the problem though.

DMaroo commented:
The issue is that, in MacOS, the arguments `termp` and `winp` are not `const` qualified (https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/openpty.3.html), but in Linux they are (https://man7.org/linux/man-pages/man3/openpty.3.html). So passing the `const void *` gives a warning in MacOS, and since `-werror` is enabled, the build fails.

DMaroo commented:
Also, I'm more concerned why the Windows Appveyor job fails. It even fails to compile which seems very surprising. I hope it is not some MSVC specific thing (or worse, a MSVC bug).

DMaroo commented:
As for NetBSD, I will try researching it more. It should not really fail since Linux and OpenBSD have the same API for the PTY family of functions.

DMaroo committed:
Remove const qualifiers for termios and winsize arguments

  * This makes the build pass on MacOS, which does accepts non-const
    arguments

DMaroo commented:
Now the compilations finished successfully, but the `test_subprocess` unit test fails on MacOS. I have no idea why. I believe the reason may be similar to the one for NetBSD, but I don't know why.

@ret2libc It would be nice if you could also run the ninja tests on your MacOS machine. Thanks.

ret2libc commented:
> Now the compilations finished successfully, but the `test_subprocess` unit test fails on MacOS. I have no idea why. I believe the reason may be similar to the one for NetBSD, but I don't know why.
> 
> @ret2libc It would be nice if you could also run the ninja tests on your MacOS machine. Thanks.

I see MacOS working well now. Where is the failing job?

XVilka commented:
> Also, I'm more concerned why the Windows Appveyor job fails. It even fails to compile which seems very surprising. I hope it is not some MSVC specific thing (or worse, a MSVC bug).

It has nothing to do with MSVC, `clang-cl` fails too:
```
FAILED: librz/util/rz_util-0.5.dll.p/alloc.c.obj 
"clang-cl" "-Ilibrz\util\rz_util-0.5.dll.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Ilibrz\util\sdb" "-I..\librz\util\sdb" "-Isubprojects\zlib-1.2.12" "-I..\subprojects\zlib-1.2.12" "-I..\subprojects\xz-5.2.5\src\liblzma\api" "/MD" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "-DPSAPI_VERSION=1" "-fcommon" "-D__STDC__=1" "-D_CRT_DECLARE_NONSTDC_NAMES " "-D_CRT_SECURE_NO_WARNINGS" "-D_CRT_NONSTDC_NO_DEPRECATE" "-DRZ_PLUGIN_INCORE=1" "-Werror=sizeof-pointer-memaccess" "-DLZMA_API_STATIC" "/Fdlibrz\util\rz_util-0.5.dll.p\alloc.c.pdb" /Folibrz/util/rz_util-0.5.dll.p/alloc.c.obj "/c" ../librz/util/alloc.c
In file included from ../librz/util/alloc.c:4:
In file included from ..\librz\include/rz_util.h:62:
..\librz\include/rz_util/rz_sys.h(47,8): error: unknown type name 'pid_t'
RZ_API pid_t rz_sys_forkpty(int *amaster, char *name, void /* const struct termios */ *termp, void /* const struct winsize */ *winp);
       ^
1 error generated.
```

DMaroo committed:
Change `pid_t` to int so that it compiles on Windows

DMaroo commented:
> It has nothing to do with MSVC, `clang-cl` fails too:
> 
> ```
> FAILED: librz/util/rz_util-0.5.dll.p/alloc.c.obj 
> "clang-cl" "-Ilibrz\util\rz_util-0.5.dll.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Ilibrz\util\sdb" "-I..\librz\util\sdb" "-Isubprojects\zlib-1.2.12" "-I..\subprojects\zlib-1.2.12" "-I..\subprojects\xz-5.2.5\src\liblzma\api" "/MD" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "-DPSAPI_VERSION=1" "-fcommon" "-D__STDC__=1" "-D_CRT_DECLARE_NONSTDC_NAMES " "-D_CRT_SECURE_NO_WARNINGS" "-D_CRT_NONSTDC_NO_DEPRECATE" "-DRZ_PLUGIN_INCORE=1" "-Werror=sizeof-pointer-memaccess" "-DLZMA_API_STATIC" "/Fdlibrz\util\rz_util-0.5.dll.p\alloc.c.pdb" /Folibrz/util/rz_util-0.5.dll.p/alloc.c.obj "/c" ../librz/util/alloc.c
> In file included from ../librz/util/alloc.c:4:
> In file included from ..\librz\include/rz_util.h:62:
> ..\librz\include/rz_util/rz_sys.h(47,8): error: unknown type name 'pid_t'
> RZ_API pid_t rz_sys_forkpty(int *amaster, char *name, void /* const struct termios */ *termp, void /* const struct winsize */ *winp);
>        ^
> 1 error generated.
> ```

~~I have added a preprocessor define for this in `rz_sys.h`. The Windows CI should now not fail because of this.~~

I have changed the function signature for `rz_sys_forkpty` to return an `int` instead of `pid_t`. This will (hopefully) allow compilation on Windows.

XVilka commented:
Windows still has a few compilation problems with `clang-cl`:
```
"clang-cl" "-Ilibrz\util\rz_util-0.5.dll.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Ilibrz\util\sdb" "-I..\librz\util\sdb" "-Isubprojects\zlib-1.2.12" "-I..\subprojects\zlib-1.2.12" "-I..\subprojects\xz-5.2.5\src\liblzma\api" "/MD" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "-DPSAPI_VERSION=1" "-fcommon" "-D__STDC__=1" "-D_CRT_DECLARE_NONSTDC_NAMES " "-D_CRT_SECURE_NO_WARNINGS" "-D_CRT_NONSTDC_NO_DEPRECATE" "-DRZ_PLUGIN_INCORE=1" "-Werror=sizeof-pointer-memaccess" "-DLZMA_API_STATIC" "/Fdlibrz\util\rz_util-0.5.dll.p\subprocess.c.pdb" /Folibrz/util/rz_util-0.5.dll.p/subprocess.c.obj "/c" ../librz/util/subprocess.c
../librz/util/subprocess.c(81,44): warning: incompatible pointer types passing 'RzThreadLock **' (aka 'struct rz_th_lock_t **') to parameter of type 'void *volatile *' [-Wincompatible-pointer-types]
                lock = InterlockedCompareExchangePointer(&subproc_mutex, INVALID_POINTER_VALUE, INVALID_POINTER_VALUE);
                                                         ^~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um\winnt.h(3219,43): note: passing argument to parameter 'Destination' here
    _Interlocked_operand_ PVOID volatile *Destination,
                                          ^
../librz/util/subprocess.c(92,31): warning: incompatible pointer types passing 'RzThreadLock **' (aka 'struct rz_th_lock_t **') to parameter of type 'void *volatile *' [-Wincompatible-pointer-types]
                        InterlockedExchangePointer(&subproc_mutex, INVALID_POINTER_VALUE);
                                                   ^~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um\winnt.h(3229,43): note: passing argument to parameter 'Target' here
    _Interlocked_operand_ PVOID volatile *Target,
                                          ^
../librz/util/subprocess.c(98,30): warning: incompatible pointer types passing 'RzThreadLock **' (aka 'struct rz_th_lock_t **') to parameter of type 'void *volatile *' [-Wincompatible-pointer-types]
                InterlockedExchangePointer(&subproc_mutex, lock);
                                           ^~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um\winnt.h(3229,43): note: passing argument to parameter 'Target' here
    _Interlocked_operand_ PVOID volatile *Target,
                                          ^
../librz/util/subprocess.c(147,37): warning: incompatible pointer types passing 'RzThreadLock **' (aka 'struct rz_th_lock_t **') to parameter of type 'void *volatile *' [-Wincompatible-pointer-types]
                lock = InterlockedExchangePointer(&subproc_mutex, NULL);
                                                  ^~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um\winnt.h(3229,43): note: passing argument to parameter 'Target' here
    _Interlocked_operand_ PVOID volatile *Target,
                                          ^
../librz/util/subprocess.c(150,30): warning: incompatible pointer types passing 'RzThreadLock **' (aka 'struct rz_th_lock_t **') to parameter of type 'void *volatile *' [-Wincompatible-pointer-types]
                InterlockedExchangePointer(&subproc_mutex, lock);
                                           ^~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um\winnt.h(3229,43): note: passing argument to parameter 'Target' here
    _Interlocked_operand_ PVOID volatile *Target,
                                          ^
../librz/util/subprocess.c(249,22): error: conflicting types for 'rz_subprocess_start_opt'
RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
                     ^
..\librz\include/rz_util/rz_subprocess.h(102,22): note: previous declaration is here
RZ_API RzSubprocess *rz_subprocess_start_opt(RZ_NONNULL const RzSubprocessOpt *opt);
                     ^
5 warnings and 1 error generated.
[376/1967] Compiling C object librz/util/rz_util-0.5.dll.p/sys.c.obj
../librz/util/sys.c(302,9): warning: TODO: rz_sys_bt : unimplemented [-W#pragma-messages]
#pragma message("TODO: rz_sys_bt : unimplemented")
        ^
../librz/util/sys.c(487,9): warning: rz_sys_crash_handler : unimplemented for this platform [-W#pragma-messages]
#pragma message("rz_sys_crash_handler : unimplemented for this platform")
        ^
../librz/util/sys.c(673,9): warning: rz_sys_cmdbg is not implemented for this platform [-W#pragma-messages]
#pragma message("rz_sys_cmdbg is not implemented for this platform")
        ^
../librz/util/sys.c(1150,12): warning: unused variable 'val_utf16' [-Wunused-variable]
                wchar_t *val_utf16 = NULL;
                         ^
4 warnings generated.
```
And with Visual Studio 2022 MSVC:
```
"cl" "-Ilibrz\socket\rz_socket.lib.p" "-I." "-I.." "-Ilibrz" "-I..\librz" "-Ilibrz\include" "-I..\librz\include" "-Ilibrz\util\sdb\src" "-I..\librz\util\sdb\src" "-Ilibrz\util\sdb" "-I..\librz\util\sdb" "/MT" "/nologo" "/showIncludes" "/utf-8" "/W2" "/O2" "/Gw" "-DPSAPI_VERSION=1" "-DRZ_PLUGIN_INCORE=1" "/Fdlibrz\socket\rz_socket.lib.p\run.c.pdb" /Folibrz/socket/rz_socket.lib.p/run.c.obj "/c" ../librz/socket/run.c
TODO: handle_redirection_proc: Not implemented for this platform
string redirection handle not yet done
../librz/socket/run.c(791): warning C4311: 'type cast': pointer truncation from 'void *' to 'int'
../librz/socket/run.c(958): error C2065: 'pid_t': undeclared identifier
../librz/socket/run.c(958): error C2146: syntax error: missing ';' before identifier 'child_pid'
../librz/socket/run.c(958): error C2065: 'child_pid': undeclared identifier
../librz/socket/run.c(959): error C2065: 'child_pid': undeclared identifier
../librz/socket/run.c(964): error C2065: 'child_pid': undeclared identifier
../librz/socket/run.c(1139): warning C4312: 'type cast': conversion from 'int' to 'void *' of greater size
```

DMaroo committed:
Add empty implementation for PTY API for Windows

  * Update function signature for `rz_subprocess_start_opt` for Windows
  * Re-add `#define` for `pid_t` in `librz/socket/run.c`

DMaroo committed:
Move functions to remove duplicate definitions

DMaroo commented:
I don't know why the `macos-meson-clang-tests` CI is failing. Can someone run and check it locally?

DMaroo commented:
Something weird is happening on MacOS platforms. The tests seem to pass sometimes, and fail sometimes. I don't know why there's this sort of inconsistency.

ret2libc commented:
I'm going to try it later tonight! I'll test with ASAN/UBSAN as well.

ret2libc commented:
@DMaroo I did this:
```
diff --git a/test/unit/test_subprocess.c b/test/unit/test_subprocess.c
index 1c2409314b..0469a2fda7 100644
--- a/test/unit/test_subprocess.c
+++ b/test/unit/test_subprocess.c
@@ -275,6 +275,7 @@ bool test_interactive_pty(void) {

        RzStrBuf *sb = rz_subprocess_stdout_readline(sp, UT_TIMEOUT);
        int c = atoi(rz_strbuf_get(sb));
+       eprintf("c = %d\n", c);
        char buf[100];
        snprintf(buf, sizeof(buf), "%d\n", 3 + 5 + c);
        rz_subprocess_stdin_write(sp, (const ut8 *)buf, strlen(buf));
@@ -334,17 +335,17 @@ bool test_interactive_custom_pty(void) {
 #endif // PTY functions
```
and also
```
@@ -1141,6 +1136,7 @@ static size_t read_to_strbuf(RzStrBuf *sb, int fd, bool *fd_eof, size_t n_bytes,
                to_read = n_bytes;
        }
        ssize_t sz = read(fd, buf, to_read);
+       printf("buf = %*s\n", sz, buf);
        if (sz == 0 || (is_pty && sz == -1 && errno == EIO)) {
                /* In case of PTY, EIO (input/output error) denotes EOF, hence the manual checking */
                *fd_eof = true;
```
and you can see that `c` is not correctly computed and that buf, at least in my case, has some things from stderr. Indeed if i execute the subprocess-interactive program I get the first line `subprocess-interactive(25345,0x10d292600) malloc: nano zone abandoned due to inability to preallocate reserved vm space.`

I *believe* (but you are now probably more into this than I am right now, so I may miss some important details) we should not use the `forkpty` but `openpty` + handling the master_fd so that we can split it into stdout/stderr/stdin according to `opt->std*_pipe` values and then do a regular `fork` (no `login_tty`, which seems to be the one merging in/out/err into one thing). Does this make sense?

DMaroo commented:
> I _believe_ (but you are now probably more into this than I am right now, so I may miss some important details) we should not use the `forkpty` but `openpty` + handling the master_fd so that we can split it into stdout/stderr/stdin according to `opt->std*_pipe` values and then do a regular `fork` (no `login_tty`, which seems to be the one merging in/out/err into one thing). Does this make sense?

Oh, so we dup the stream to `master_fd` and `slave_fd` (for the parent and child, respectively) if the pipe option for that stream has been set to `RZ_SUBPROCESS_PIPE_CREATE`. Is that what you meant? @ret2libc

ret2libc commented:
> > I _believe_ (but you are now probably more into this than I am right now, so I may miss some important details) we should not use the `forkpty` but `openpty` + handling the master_fd so that we can split it into stdout/stderr/stdin according to `opt->std*_pipe` values and then do a regular `fork` (no `login_tty`, which seems to be the one merging in/out/err into one thing). Does this make sense?
> 
> Oh, so we dup the stream to `master_fd` and `slave_fd` (for the parent and child, respectively) if the pipe option for that stream has been set to `RZ_SUBPROCESS_PIPE_CREATE`. Is that what you meant? @ret2libc

Not sure about `slave_fd`, but I think `master_fd` should be duplicated to be stdout/stderr/stdin only according to the `opt->std*_pipe` options. If `opt->stderr_pipe` is PIPE_NONE, you don't make it make `master_fd` the duplicate of `2`. If `opt->stdout_pipe` is PIPE_CREATE you duplicate `master_fd` to be `1`. I think this helps.

DMaroo committed:
Allow specifying piping options even in `forkpty` mode

    * Only expose the master fd for the streams which have pipe create
      as their piping option

DMaroo commented:
> Not sure about slave_fd, but I think master_fd should be duplicated to be stdout/stderr/stdin only according to the opt->std*_pipe options. If opt->stderr_pipe is PIPE_NONE, you don't make it make master_fd the duplicate of 2. If opt->stdout_pipe is PIPE_CREATE you duplicate master_fd to be 1. I think this helps.

Understood. I have first obtained a pair of master-slave file descriptors through `openpty` (in case no `RzPTY` argument passed was `NULL`). I then only assign the `master_fd` to the streams which have `RZ_SUBPROCESS_PIPE_CREATE` as their pipe mode. This way, there is no way for the user (using the API) to access any of the streams blocked out, unless they go and manually access the `master_fd`. Then I fork normally, and use `rz_login_tty` for the child process.

DMaroo committed:
Add comment about not using `master_fd` directly

ret2libc commented:
@DMaroo I think you broke something with the last changes

DMaroo committed:
Fix types in definition for Windows

DMaroo committed:
Free the allocated `RzPTY` object

DMaroo committed:
Add const qualifiers for `RzPTY` argument

DMaroo commented:
~~Many of the tests are failing because of completely unrelated reasons. I have no idea why.~~

Alright, my bad. Changes have broken `rz-test` (since it uses the subprocess API). I'm looking into it.

Update: Should work now!

DMaroo committed:
Assign `fork_mode` and `pty` in all instances of subprocess opt start

    * for good measure

DMaroo committed:
Compare master fd with -1, not 0

DMaroo committed:
Remove stray file

DMaroo commented:
Only the OpenBSD job is failing now. I don't know how to reproduce the failure to fix it.

ret2libc commented on librz/util/sys.c:
```diff
@@ -603,6 +615,8 @@ RZ_API int rz_sys_cmd_str_full(const char *cmd, const char *input, char **output
 		.stdin_pipe = input ? RZ_SUBPROCESS_PIPE_CREATE : RZ_SUBPROCESS_PIPE_NONE,
 		.stdout_pipe = output ? RZ_SUBPROCESS_PIPE_CREATE : RZ_SUBPROCESS_PIPE_NONE,
 		.stderr_pipe = sterr ? RZ_SUBPROCESS_PIPE_CREATE : RZ_SUBPROCESS_PIPE_NONE,
+		.fork_mode = RZ_SUBPROCESS_FORK,
```
I see there are many of these cases. Now, or in the future, it might make sense to create a `rz_subprocess_opt_init` to avoid this and prevent errors from uninitialized fields.

ret2libc replied:
If not now, please create an issue for this.

DMaroo replied:
To be fair, `RZ_SUBPROCESS_FORK` has the value 0. So even if we don't initialize the `fork_mode` and `pty` field explicitly, we would still get `fork_mode = RZ_SUBPROCESS_FORK` and `pty = NULL`. I just explicitly put them everywhere for the sake of it.

DMaroo replied:
Creating an `rz_subprocess_opt_init` does not make sense to me because all it would do is just take every argument and add it to a `RzSubprocessOpt` struct. I would rather have struct initialization than a 9 argument function (which would be very ugly)>

ret2libc replied:
I didn't mean to put all fields as args. What we have done for similar cases is to put common options as arguments, and then the caller will setup the extra/special fields. By doing so you can ensure all fields are initialized correctly.

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -927,53 +985,67 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+
+		if (pty) {
+			proc->master_fd = master_fd = pty->master_fd;
+			if (tcsetattr(pty->slave_fd, TCSANOW, term_params) == -1) {
+				perror("tcsetattr");
```
what is this? A comment like what's done above would help understand.

DMaroo replied:
Ah, sure. I'll add a comment.

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -995,6 +1068,16 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+			if (close(master_fd) == -1) {
+				perror("close");
+			}
+			if (rz_sys_login_tty(pty->slave_fd) == -1) {
+				perror("login_tty");
+				rz_sys_exit(-1, true);
+			}
```
Is this really needed? If yes, why? Maybe a comment.

DMaroo replied:
![pty](https://user-images.githubusercontent.com/72350048/208911158-73cfe1c5-6dfc-4921-9d61-4ec1b4894d0c.jpg)

As labelled in the above chart, `login_tty` is responsible for connecting the child process' file descriptor to the slave file descriptor.

ret2libc replied:
Awesome 👍

ret2libc commented:
https://docs.pwntools.com/en/stable/tubes/processes.html is very interesting. I think what I initially wanted to do (loooong time ago) was to mimick something like this. Maybe we should take a similar approach and use `RzSubprocessPipeCreate` to specify whether a PTY is requested or not for stdin/stdout/stderr.

ret2libc commented:
Also, at least as a local test (not to push) it would be nice to run `python` with the new settings and see whether in one case it prints `>>> xxx` an in the other it doesn't. `python` will behave differently based on whether it's connected to PTY or not.

Another example would be to write an auxiliary program that uses libc stdout buffering and see if the new settings we can circument this (this was essentially the very main reason to do this).

DMaroo committed:
Add second comment about the terminal attribute

DMaroo commented:
@ret2libc

After going through `pwntools`'s source, I agree with your suggestion. It seems much more clearer. Thanks!

DMaroo commented:
Also, `pwntools`'s API has an argument `raw`, which is used to specify whether the opened PTY should be in raw mode or not. It is `True` by default, and I also directly make the PTY raw in case of `RzSubprocess`. Should I also add such an argument for our API?

ret2libc commented:
I think we can skip it for now, at least in this PR.

DMaroo committed:
Overhaul the whole PTY implementation

    * Now the caller can either use a pipe or PTY to send and receive
      data from any of the streams
    * Extra option `make_raw` to use the slave PTY in raw mode
    * Add documentation for all the new functions

DMaroo commented:
> I think we can skip it for now, at least in this PR.

Oh, I'm sorry. I didn't read this, but I have added a `make_raw` option. It's only a single if-condition.

DMaroo committed:
Fix UAF bug

DMaroo commented:
Again, the  test results are quite unexpected. MacOS arm64 passes, but MacOS clang tests doesn't (lol).

ret2libc commented on librz/util/subprocess.c:
```diff
@@ -890,10 +902,103 @@ static void destroy_child_env(char **child_env) {
+		proc->slave_fd = pty->slave_fd;
+	}
+
+	switch (opt->stdin_pipe) {
+	case RZ_SUBPROCESS_PIPE_CREATE:
+		if (rz_sys_pipe(stdin_pipe, true) == -1) {
+			perror("pipe");
+			return false;
+		}
+		proc->stdin_fd = stdin_pipe[1];
+		break;
+	case RZ_SUBPROCESS_PIPE_PTY:
+		stdin_pipe[0] = pty->slave_fd;
+		stdin_pipe[1] = pty->master_fd;
+		proc->stdin_fd = stdin_pipe[1];
+		break;
+	case RZ_SUBPROCESS_PIPE_STDOUT:
+		RZ_LOG_ERROR("Invalid pipe option 'RZ_SUBPROCESS_PIPE_STDOUT' for stdin. Ignoring...\n");
+		break;
+	case RZ_SUBPROCESS_PIPE_NONE:
+		break;
+	}
```
extract in a function

DMaroo replied:
I think is better the way it is right now. Extracting it to a function makes the code very convoluted and contrived since in every case, we are doing something different. Also, the reduction in code size, if we use functions, is not significant.

DMaroo committed:
Add error handling for `tcgetattr`

DMaroo committed:
Move the term param setting code

XVilka commented on librz/util/subprocess.c:
```diff
@@ -246,7 +247,7 @@ static LPWCH override_env(const char *envvars[], const char *envvals[], size_t e
 }
 
-RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
+RZ_API RzSubprocess *rz_subprocess_start_opt(RZ_NONNULL const RzSubprocessOpt *opt) {
```
Please add Doxygen

XVilka commented on librz/util/subprocess.c:
```diff
@@ -700,6 +701,15 @@ RZ_API void rz_subprocess_free(RzSubprocess *proc) {
 	free(proc);
 }
+
+RZ_API RZ_OWN RzPty *rz_subprocess_openpty(RZ_BORROW RZ_NULLABLE char *slave_name, RZ_NULLABLE void /* const struct termios */ *term_params, RZ_NULLABLE void /* const struct winsize */ *win_params) {
```
Doxygen

XVilka commented on librz/util/subprocess.c:
```diff
@@ -714,6 +724,8 @@ struct rz_subprocess_t {
 	int ret;
 	RzStrBuf out;
 	RzStrBuf err;
+	int master_fd; ///< Needed to check whether PTY
```
Maybe better to rename to something more neutral?

DMaroo replied:
I agree about it not being neutral, but it is the standard term used for it. It will be much easier to understand if I also use the same term that the POSIX documentation uses.

XVilka commented on librz/util/subprocess.c:
```diff
@@ -890,10 +902,103 @@ static void destroy_child_env(char **child_env) {
+	return true;
+}
+
+RZ_API RzSubprocess *rz_subprocess_start_opt(RZ_NONNULL const RzSubprocessOpt *opt) {
```
Doxygen and ownership mark of the return

XVilka commented on librz/util/subprocess.c:
```diff
@@ -1281,17 +1417,98 @@ RZ_API void rz_subprocess_free(RzSubprocess *proc) {
+/**
+ * \brief Call openpty(3) with the provided arguments
+ *
+ * \param slave_name The name of the slave PTY is stored in this
```
Please rename the parameter

XVilka commented on librz/util/subprocess.c:
```diff
@@ -1281,17 +1417,98 @@ RZ_API void rz_subprocess_free(RzSubprocess *proc) {
+	}
+}
+
+/**
```
This Doxygen is unnecessary and can be removed

DMaroo replied:
Agreed, but I'll let it stay.

XVilka commented on librz/util/sys.c:
```diff
@@ -74,6 +74,18 @@ extern char **environ;
 
+/* For "openpty" family of funtcions */
+#if HAVE_OPENPTY && HAVE_FORKPTY && HAVE_LOGIN_TTY
+#if defined(__APPLE__) || defined(__NetBSD__) || defined(__OpenBSD__)
```
Could it be converted to use just Meson? cc @ret2libc

XVilka commented on librz/util/sys.c:
```diff
@@ -1681,6 +1696,45 @@ RZ_API int rz_sys_fork(void) {
 }
 #endif
 
+RZ_API /* pid_t */ int rz_sys_forkpty(int *amaster, char *name, void /* const struct termios */ *termp, void /* const struct winsize */ *winp) {
```
Doxygen

XVilka commented on librz/util/sys.c:
```diff
@@ -1681,6 +1696,45 @@ RZ_API int rz_sys_fork(void) {
+#endif
+}
+
+RZ_API int rz_sys_openpty(int *amaster, int *aslave, char *name, void /* const struct termios */ *termp, void /* const struct winsize */ *winp) {
```
Doxygen

XVilka commented on librz/util/sys.c:
```diff
@@ -1681,6 +1696,45 @@ RZ_API int rz_sys_fork(void) {
+#endif
+}
+
+RZ_API int rz_sys_login_tty(int fd) {
```
Doxygen

DMaroo committed:
Add documentation for the subprocess API and sys API

DMaroo committed:
Format according to clang-format
